core/tests/protractor/learnerFlow.js

```diff
@@ -0,0 +1,415 @@
+// Copyright 2018 The Oppia Authors. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS-IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/**
+ * @fileoverview End-to-end tests for the learner flow.
+ */
+
+var forms = require('../protractor_utils/forms.js');
+var general = require('../protractor_utils/general.js');
+var users = require('../protractor_utils/users.js');
+var waitFor = require('../protractor_utils/waitFor.js');
+var workflow = require('../protractor_utils/workflow.js');
+
+var AdminPage = require('../protractor_utils/AdminPage.js');
+var CreatorDashboardPage =
+  require('../protractor_utils/CreatorDashboardPage.js');
+var CollectionEditorPage =
+  require('../protractor_utils/CollectionEditorPage.js');
+var ExplorationEditorPage =
+  require('../protractor_utils/ExplorationEditorPage.js');
+var ExplorationPlayerPage =
+  require('../protractor_utils/ExplorationPlayerPage.js');
+var LearnerDashboardPage =
+  require('../protractor_utils/LearnerDashboardPage.js');
+var LibraryPage = require('../protractor_utils/LibraryPage.js');
+
+describe('Learner dashboard functionality', function() {
+  var adminPage = null;
+  var creatorDashboardPage = null;
+  var explorationEditorPage = null;
+  var explorationEditorMainTab = null;
+  var explorationPlayerPage = null;
+  var libraryPage = null;
+  var learnerDashboardPage = null;
+  var oppiaLogo = element(by.css('.protractor-test-oppia-main-logo'));
+  var continueButton = element(by.css('.protractor-test-continue-button'));
+  var clickContinueButton = function () {
+    waitFor.elementToBeClickable(
+      continueButton, 'Could not click continue button');
+    continueButton.click();
+    waitFor.pageToFullyLoad();
+  };
+
+  var createAboutOppiaExploration = function() {
```

Comment:
Perhaps createAboutOppiaExplorationOnDesktop()?

In general, actually, would be better to have createAboutOppiaExploration() and have the is-mobile-or-not branching in there. Restrict the is-mobile branching to as small and contained a part of the codebase as possible, and for tests in protractor/, have them largely be the same in terms of what operations are done on mobile or desktop. Currently I think the branches are a bit too big, they're almost like two entirely separate tests. This comment applies throughout the PR.

Reply:
Is that the only difference (i.e. after you create the explorations, the operations on them are similar)?

If so, could we just do an if/else on the setup? E.g. `setupExploration(explorationData, demoExplorationId);` which loads the demo exploration on mobile and sets up the exploration manually on desktop?

Reply:
I think we can do something of that sort here, not sure to what extent but yes, I will try this out.

Reply:
Okay, so what I did here was: I abstracted all common operations, which basically includes navigating to the incomplete and complete sections of the learner dashboard, navigating to the incomplete and complete sections of explorations/collections and then checking whether the exploration/collection is rightly placed.
These have been put into functions which have been appropriately called then. This has helped us to do away with duplication of code and the spec now looks more unified.